java/ql/src/semmle/code/java/frameworks/apache/Collections.qll

```diff
@@ -37,10 +37,187 @@ private class ApacheCollectionsModel extends SummaryModelCsv {
+        "org.apache.commons.collections4;MultiValuedMap;true;put;;;Argument[1];Element of MapValue of Argument[-1];value",
+        "org.apache.commons.collections4;MultiValuedMap;true;putAll;(Object,Iterable);;Argument[0];MapKey of Argument[-1];value",
+        "org.apache.commons.collections4;MultiValuedMap;true;putAll;(Object,Iterable);;Element of Argument[1];Element of MapValue of Argument[-1];value",
+        "org.apache.commons.collections4;MultiValuedMap;true;putAll;(Map);;MapKey of Argument[0];MapKey of Argument[-1];value",
+        "org.apache.commons.collections4;MultiValuedMap;true;putAll;(Map);;MapValue of Argument[0];Element of MapValue of Argument[-1];value",
+        "org.apache.commons.collections4;MultiValuedMap;true;putAll;(MultiValuedMap);;MapKey of Argument[0];MapKey of Argument[-1];value",
+        "org.apache.commons.collections4;MultiValuedMap;true;putAll;(MultiValuedMap);;Element of MapValue of Argument[0];Element of MapValue of Argument[-1];value",
+        "org.apache.commons.collections4;MultiValuedMap;true;remove;;;Element of MapValue of Argument[-1];Element of ReturnValue;value",
+        "org.apache.commons.collections4;MultiValuedMap;true;values;;;Element of MapValue of Argument[-1];Element of ReturnValue;value",
+        "org.apache.commons.collections4;OrderedIterator;true;previous;;;Element of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections4;OrderedMap;true;firstKey;;;MapKey of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections4;OrderedMap;true;lastKey;;;MapKey of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections4;OrderedMap;true;nextKey;;;MapKey of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections4;OrderedMap;true;previousKey;;;MapKey of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections4;Put;true;put;;;MapValue of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections4;Put;true;put;;;Argument[0];MapKey of Argument[-1];value",
+        "org.apache.commons.collections4;Put;true;put;;;Argument[1];MapValue of Argument[-1];value",
+        "org.apache.commons.collections4;Put;true;putAll;(Map);;MapKey of Argument[0];MapKey of Argument[-1];value",
+        "org.apache.commons.collections4;Put;true;putAll;(Map);;MapValue of Argument[0];MapValue of Argument[-1];value",
+        "org.apache.commons.collections4;SortedBag;true;first;;;Element of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections4;SortedBag;true;last;;;Element of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections4;Trie;true;prefixMap;;;MapKey of Argument[-1];MapKey of ReturnValue;value",
+        "org.apache.commons.collections4;Trie;true;prefixMap;;;MapValue of Argument[-1];MapValue of ReturnValue;value",
+        "org.apache.commons.collections;ArrayStack;true;peek;;;Element of Argument[-1];ReturnValue;value",
```

Comment:
Only reviewed the `collections4` models. Are the two sets of models identical or supposed to be? If so use something like `"org.apache.commons.collections" + ["", "4"] + ";" + ["ArrayStack;true;peek;;;Element of Argument[-1];ReturnValue;value", ...]`

This way we can at least guard against the models drifting apart by accident.

Reply:
My review assumed both sets were identical too. +1 to this suggestion.